plugins/processors/reversedns/rdnscache.go

```diff
@@ -0,0 +1,324 @@
+	return d
+}
+
+// dnslookup represents a lookup request/response. It may or may not be answered yet.
+// interested parties register themselves with existing requests or create new ones
+// to get their dns query answered. Answers will be pushed out to callbacks.
+type dnslookup struct {
+	ip        string // keep a copy for the expireList.
+	domains   []string
+	expiresAt time.Time
+	completed bool
+	callbacks []callbackChannelType
+}
+
+type callbackChannelType chan []string
+
+// Lookup takes a string representing a parseable ipv4 or ipv6 IP, and blocks
+// until it has resolved to 0-n results, or until its lookup timeout has elapsed.
+// if the lookup timeout elapses, it returns an empty slice.
+func (d *ReverseDNSCache) Lookup(ip string) []string {
+	if len(ip) == 0 {
+		return []string{}
+	}
+	// todo: fixed response for known localhost, etc.
+	return d.lookup(ip)
+}
+
+func (d *ReverseDNSCache) lookup(ip string) []string {
+	// check if the value is cached
+	d.rwLock.RLock()
+	result, found := d.unsafeGetFromCache(ip)
+	if found && result.completed && result.expiresAt.After(time.Now()) {
+		debug("found in cache", result)
+		defer d.rwLock.RUnlock()
+		atomic.AddUint64(&d.stats.CacheHit, 1)
+		// cache is valid
+		debug("Found in cache: ", result.domains)
+		return result.domains
+	}
+	d.rwLock.RUnlock()
+
+	// if it's not cached, kick off a lookup job and subscribe to the result.
+	debug("subscribe!")
+	lookupChan := d.subscribeTo(ip)
+	timer := time.NewTimer(d.lookupTimeout)
+	defer timer.Stop()
+
+	// timer is still necessary even if doLookup respects timeout due to worker
+	// pool starvation.
+	select {
+	case domains := <-lookupChan:
+		return domains
+	case <-timer.C:
+		fmt.Fprintf(os.Stderr, "reverse dns lookup timed out\n")
+		return []string{}
+	}
+}
+
+func (d *ReverseDNSCache) subscribeTo(ip string) callbackChannelType {
+	callback := make(callbackChannelType, 1)
+
+	d.rwLock.Lock()
+	defer d.rwLock.Unlock()
+	debug("d.rwLock.Lock()")
+	defer debug("d.rwLock.Unlock()")
+
+	// confirm it's still not in the cache. This needs to be done under an active lock.
+	result, found := d.unsafeGetFromCache(ip)
+	if found {
+		debug("found")
+		atomic.AddUint64(&d.stats.CacheHit, 1)
+		// has the request been answered since we last checked?
+		if result.completed {
+			debug("found completed")
+			// we can return the answer with the channel.
+			callback <- result.domains
+			return callback
+		}
+		debug("found not completed")
+		// there's a request but it hasn't been answered yet;
+		// add yourself to the subscribers and return that.
+		result.callbacks = append(result.callbacks, callback)
+		d.unsafeSaveToCache(result)
+		return callback
+	}
+
+	atomic.AddUint64(&d.stats.CacheMiss, 1)
+
+	debug("not found")
+
+	// otherwise we need to register the request
+	l := &dnslookup{
+		ip:        ip,
+		expiresAt: time.Now().Add(d.ttl),
+		callbacks: []callbackChannelType{callback},
+	}
+
+	d.unsafeSaveToCache(l)
+	go d.doLookup(*l)
```

Comment:
unsafeSaveToCache wrote it to the cache and we do not want to share copies to it because we no longer "own" it. We have to drop the reference to it before the function unlocks, so the goroutine needs its own copy.

Reply:
Confusing and prone to error under future modification IMO, maybe just pass in the parts you need instead of the entire struct.